File: src/Scripting/RubyRB/example2.rb
# Abstract Factory
class DatabaseFactory
    def create_database
      raise NotImplementedError, 'create_database() must be implemented'
    end
  end
  
  # Concrete Factories
  class PostgresFactory < DatabaseFactory
    def create_database
      Postgres.new
    end
  end
  
  class MySQLFactory < DatabaseFactory
    def create_database
      MySQL.new
    end
  end
  
  # Products
  class Postgres
    def connect
      puts 'Connecting to PostgreSQL'
    end
  
    def query
      puts 'Querying PostgreSQL'
    end
  end
  
  class MySQL
    def connect
      puts 'Connecting to MySQL'
    end
  
    def query
      puts 'Querying MySQL'
    end
  end
  
  def use_database(factory)
    db = factory.create_database
    db.connect
    db.query
  end
  
  # Usage
  use_database(PostgresFactory.new)
  use_database(MySQLFactory.new)
  

File: src/Web/JavaScriptJS/example2.js
// Abstract Factory
function createDatabase(dbType) {
    if (dbType === "postgresql") {
        return new Postgres();
    } else if (dbType === "mysql") {
        return new MySQL();
    }
}

// Concrete Products
class Postgres {
    connect() {
        console.log("Connecting to PostgreSQL");
    }

    query() {
        console.log("Querying PostgreSQL");
    }
}

class MySQL {
    connect() {
        console.log("Connecting to MySQL");
    }

    query() {
        console.log("Querying MySQL");
    }
}

// Usage
const db1 = createDatabase("postgresql");
db1.connect();
db1.query();

const db2 = createDatabase("mysql");
db2.connect();
db2.query();
